### DISCUSSION TITLE
Is there a PostProcessor/Output object that can show a variable value along a boundary?

### DISCUSSION THREAD

**Discussion by richmondodufisan** [idaholab/moose#27232] [Q&A General] [answered]
Hi,

I have a problem with a periodic solution. 

![Screenshot 2024-03-29 004007](https://github.com/idaholab/moose/assets/52802063/28abc535-0106-4b02-bed9-ddde0176d233)

When I get the simulation output, my current workflow is to take the exodus file to paraview and draw a line over the mesh to get a plot of the result, which is harmonic, and then fit it to a decaying sine function to get the properties I need. I would like to automate this process as I want to run this simulation several times for various material properties. 

Is there a postprocessor object that can output the results along a boundary in excel? I browsed through the postprocessor list and I'm not sure I could find one. The only option I could think of is to use PointValue, but that would require selecting hundreds of points to keep the resolution of the sine wave good. Is there any other option for doing this?

**Comment by GiudGiud** [comment, marked as answer]
Hello

You could specify a line along the boundary if it is straight
https://mooseframework.inl.gov/source/vectorpostprocessors/LineValueSampler.html
this can also work
https://mooseframework.inl.gov/source/vectorpostprocessors/PositionsFunctorValueSampler.html
but you ll need to create a Positions object for the boundary. I d like that in MOOSE btw

If not you could make a new one, a BoundaryValueSampler

**Reply by richmondodufisan** [reply]
This should work, thanks!!

**Reply by richmondodufisan** [reply]
I liked LineValueSampler, but when I switched to a quasistatic simulation and added the 'execute_on' command (set it to final), it didn't listen and instead returned empty csv files for every timestep. I then manually set it to the default, i.e 'timestep_end' and it still returned empty csv files. I tried supplying 'timestep_end' and 'final' with and without quotes.

Only after I took out 'execute_on' and let it do the default by itself (timestep_end), did it start printing out values again. How can I fix this?

**Reply by GiudGiud** [reply]
This does not seem right. Could you make a minimal example of this?

**Reply by richmondodufisan** [reply]
Yes, sure:

https://github.com/richmondodufisan/purple/blob/main/Cornea_Hyperelastic_NeoHookean/linevaluenotworking.i

That's a minimal example where I used 'final' and it returned an empty csv file. If I took that option out completely and let it use the default, it returned the line value sample at every timestep (timestep_end)

The minimal example only uses MOOSE kernels and I used GeneratedMeshGenerator so you can download and run easily

**Reply by GiudGiud** [reply]
ok this is semi-expected. you need to add the FINAL execute_on to the CSV output to get a result 
```
  [csv]
    type = CSV
    execute_on = FINAL
  []
```

I ll add a little check / warning. seems like a gotcha right now

**Reply by richmondodufisan** [reply]
And this should be in the output block? does that set it globally then? like for example if I have a VectorPostProcessor that I want only at the final timestep but I have a PostProcessor that I want at every timestep, how do I do that then?

**Reply by GiudGiud** [reply]
You would have to have multiple CSV outputs objects. Or have multiple flags on one 

CSV output does not happen unless the flag is present in the outputs block

**Reply by GiudGiud** [reply]
It does have a global effect, in that it affects all the VPPs.

**Reply by richmondodufisan** [reply]
I'm sorry I am a bit confused by what you mean. As an example, say I have:

```
[Postprocessors]
  [displace_y_real]
    type = PointValue
    variable = disp_y_real
    point = '0.002 0.001 0'
  []
  
  [displace_y_imag]
    type = PointValue
    variable = disp_y_imag
    point = '0.002 0.001 0'
  []
[]

[VectorPostprocessors]
  [wave_profile]
    type = LineValueSampler
    variable = disp_y
    start_point = '0.002 ${mid_height} 0'
    end_point = '${fparse l_plate - 0.002} ${mid_height} 0'
    num_points = ${number_of_points}
    sort_by = x
    execute_on = final
  []
[]
```

If I understand correctly, I need to do:

```
[Outputs]
  print_linear_residuals = false
  [csv]
    type = CSV
    execute_on = FINAL
  []
  exodus = true
  [pgraph]
    type = PerfGraphOutput
    execute_on = 'final'  # Default is "final"
    level = 1             # Default is 1
  []
[]
```

for the VectorPostProcessor to work, right? What about the PostProcessor?

**Reply by GiudGiud** [reply]
postprocessor will only output on final if you do this
you could do this
```
[Outputs]
  print_linear_residuals = false
  [csv]
    type = CSV
    execute_on = 'TIMESTEP_END FINAL'
  []

**Reply by richmondodufisan** [reply]
okay thank you

**Reply by richmondodufisan** [reply]
SideValueSampler also does this if I predefine the boundary in my mesh!
